pkg/provisioners/channel_util_test.go

```diff
@@ -0,0 +1,198 @@
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/scheme"
+	runtimeClient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	eventingv1alpha1 "github.com/knative/eventing/pkg/apis/eventing/v1alpha1"
+)
+
+const (
+	channelName = "test-channel"
+	testNS      = "test-namespace"
+)
+
+var (
+	truePointer = true
+)
+
+func init() {
+	// Add types to scheme.
+	istiov1alpha3.AddToScheme(scheme.Scheme)
+	eventingv1alpha1.AddToScheme(scheme.Scheme)
+}
+
+func TestCreateK8sService(t *testing.T) {
+	want := makeK8sService()
+	client := fake.NewFakeClient()
+	got, _ := CreateK8sService(context.TODO(), client, getNewChannel())
+
+	ignore := cmpopts.IgnoreTypes(apis.VolatileTime{})
+	if diff := cmp.Diff(want, got, ignore); diff != "" {
+		t.Errorf("Service (-want, +got) = %v", diff)
+	}
+}
+
+func TestCreateK8sService_Existing(t *testing.T) {
+	want := makeK8sService()
+	client := fake.NewFakeClient(want)
+	got, _ := CreateK8sService(context.TODO(), client, getNewChannel())
+
+	ignore := cmpopts.IgnoreTypes(apis.VolatileTime{})
+	if diff := cmp.Diff(want, got, ignore); diff != "" {
+		t.Errorf("Service (-want, +got) = %v", diff)
+	}
+}
+
+func TestCreateVirtualService(t *testing.T) {
+	want := makeVirtualService()
+	client := fake.NewFakeClient()
+	got, _ := CreateVirtualService(context.TODO(), client, getNewChannel())
+
+	ignore := cmpopts.IgnoreTypes(apis.VolatileTime{})
+	if diff := cmp.Diff(want, got, ignore); diff != "" {
+		t.Errorf("VirtualService (-want, +got) = %v", diff)
+	}
+}
+
+func TestCreateVirtualService_Existing(t *testing.T) {
+	want := makeVirtualService()
+	client := fake.NewFakeClient(want)
+	got, _ := CreateVirtualService(context.TODO(), client, getNewChannel())
+
+	ignore := cmpopts.IgnoreTypes(apis.VolatileTime{})
+	if diff := cmp.Diff(want, got, ignore); diff != "" {
+		t.Errorf("VirtualService (-want, +got) = %v", diff)
+	}
+}
+
+func TestUpdateChannel(t *testing.T) {
+	oldChannel := getNewChannel()
+	client := fake.NewFakeClient(oldChannel)
+
+	want := getNewChannel()
+	AddFinalizer(want, "test-finalizer")
+	want.Status.SetSinkable("test-domain")
+	UpdateChannel(context.TODO(), client, want)
+
+	got := &eventingv1alpha1.Channel{}
+	client.Get(context.TODO(), runtimeClient.ObjectKey{Namespace: testNS, Name: channelName}, got)
+
+	ignore := cmpopts.IgnoreTypes(apis.VolatileTime{})
+	if diff := cmp.Diff(want, got, ignore); diff != "" {
+		t.Errorf("Channel (-want, +got) = %v", diff)
+	}
+}
+
+func getNewChannel() *eventingv1alpha1.Channel {
+	channel := &eventingv1alpha1.Channel{
+		TypeMeta:   channelType(),
+		ObjectMeta: om(testNS, channelName),
+		Spec: eventingv1alpha1.ChannelSpec{
+			Provisioner: &corev1.ObjectReference{
+				Name:       clusterChannelProvisionerName,
+				Kind:       "ClusterProvisioner",
```

Comment:
ClusterChannelProvisioner

Reply:
Thanks, I will address in the follow-up.